### PR TITLE
Use a deterministic snapshot name when cloning from a volume

### PR DESCRIPTION
**Pull request by ejweber**
longhorn/longhorn#5986

Before this change, the volume reconciles twice while cloning (the first reconciliation fails because the engine object used is stale). Both times, we create a snapshot, leaving us with one more snapshot than we want/need.

```
time="2023-05-31T14:55:30Z" level=info msg="Event(v1.ObjectReference{Kind:\"Volume\", Namespace:\"longhorn-system\", Name:\"pvc-c2e5f072-1002-4f7c-b300-55b7417a0839\", UID:\"3bb6feee-3fa6-4fae-ade3-9e96b4e9edf3\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"34673568\", FieldPath:\"\"}): type: 'Normal' reason: 'Degraded' volume pvc-c2e5f072-1002-4f7c-b300-55b7417a0839 became degraded"
time="2023-05-31T14:55:30Z" level=debug msg="Created snapshot fe34b167-0d67-41f0-a435-8ca081ba09e6 with labels map[longhorn.io/for-cloning-volume:pvc-c2e5f072-1002-4f7c-b300-55b7417a0839] for volume pvc-e63d887a-0bc8-45d2-b49f-fddf843aaf56"
time="2023-05-31T14:55:30Z" level=info msg="Event(v1.ObjectReference{Kind:\"Volume\", Namespace:\"longhorn-system\", Name:\"pvc-c2e5f072-1002-4f7c-b300-55b7417a0839\", UID:\"3bb6feee-3fa6-4fae-ade3-9e96b4e9edf3\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"34673568\", FieldPath:\"\"}): type: 'Normal' reason: 'VolumeCloneInitiated' source volume pvc-e63d887a-0bc8-45d2-b49f-fddf843aaf56, snapshot fe34b167-0d67-41f0-a435-8ca081ba09e6"
time="2023-05-31T14:55:30Z" level=debug msg="Requeue volume due to error <nil> or Operation cannot be fulfilled on engines.longhorn.io \"pvc-c2e5f072-1002-4f7c-b300-55b7417a0839-e-12c64c7c\": the object has been modified; please apply your changes to the latest version and try again" accessMode=rwo controller=longhorn-volume frontend=blockdev migratable=false node=eweber-v124-worker-1ae51dbb-4pngn owner=eweber-v124-worker-1ae51dbb-4pngn state=attached volume=pvc-c2e5f072-1002-4f7c-b300-55b7417a0839
time="2023-05-31T14:55:30Z" level=info msg="Event(v1.ObjectReference{Kind:\"Volume\", Namespace:\"longhorn-system\", Name:\"pvc-c2e5f072-1002-4f7c-b300-55b7417a0839\", UID:\"3bb6feee-3fa6-4fae-ade3-9e96b4e9edf3\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"34673568\", FieldPath:\"\"}): type: 'Normal' reason: 'Degraded' volume pvc-c2e5f072-1002-4f7c-b300-55b7417a0839 became degraded"
time="2023-05-31T14:55:30Z" level=debug msg="Created snapshot 208cf85c-86b4-4c91-af0f-e5fba1561379 with labels map[longhorn.io/for-cloning-volume:pvc-c2e5f072-1002-4f7c-b300-55b7417a0839] for volume pvc-e63d887a-0bc8-45d2-b49f-fddf843aaf56"
time="2023-05-31T14:55:30Z" level=info msg="Event(v1.ObjectReference{Kind:\"Volume\", Namespace:\"longhorn-system\", Name:\"pvc-c2e5f072-1002-4f7c-b300-55b7417a0839\", UID:\"3bb6feee-3fa6-4fae-ade3-9e96b4e9edf3\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"34673568\", FieldPath:\"\"}): type: 'Normal' reason: 'VolumeCloneInitiated' source volume pvc-e63d887a-0bc8-45d2-b49f-fddf843aaf56, snapshot 208cf85c-86b4-4c91-af0f-e5fba1561379"
```

After this change, the volume reconciles twice while cloning (the first reconciliation fails because the engine object used is stale). The second time, we check whether a snapshot with the deterministic name we used in the first reconciliation exists and don't create a new one.

```
time="2023-06-01T15:33:34Z" level=info msg="Event(v1.ObjectReference{Kind:\"Volume\", Namespace:\"longhorn-system\", Name:\"pvc-38791991-2471-4836-9a6d-4f2356a25421\", UID:\"f3a6a76b-a930-4973-8dde-c8ff994432c7\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"35366405\", FieldPath:\"\"}): type: 'Normal' reason: 'Degraded' volume pvc-38791991-2471-4836-9a6d-4f2356a25421 became degraded"
time="2023-06-01T15:33:34Z" level=debug msg="Created snapshot 19ff6078-ab9f-5590-9c12-9e3e25e1ff56 with labels map[longhorn.io/for-cloning-volume:pvc-38791991-2471-4836-9a6d-4f2356a25421] for volume pvc-5c5c2f22-8f1d-4e92-89af-dd40b25fc438"
time="2023-06-01T15:33:34Z" level=info msg="Event(v1.ObjectReference{Kind:\"Volume\", Namespace:\"longhorn-system\", Name:\"pvc-38791991-2471-4836-9a6d-4f2356a25421\", UID:\"f3a6a76b-a930-4973-8dde-c8ff994432c7\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"35366405\", FieldPath:\"\"}): type: 'Normal' reason: 'VolumeCloneInitiated' source volume pvc-5c5c2f22-8f1d-4e92-89af-dd40b25fc438, snapshot 19ff6078-ab9f-5590-9c12-9e3e25e1ff56"
time="2023-06-01T15:33:34Z" level=debug msg="Requeue volume due to error <nil> or Operation cannot be fulfilled on engines.longhorn.io \"pvc-38791991-2471-4836-9a6d-4f2356a25421-e-d2ed216b\": the object has been modified; please apply your changes to the latest version and try again" accessMode=rwo controller=longhorn-volume frontend=blockdev migratable=false node=eweber-v124-worker-1ae51dbb-4pngn owner=eweber-v124-worker-1ae51dbb-4pngn state=attached volume=pvc-38791991-2471-4836-9a6d-4f2356a25421
time="2023-06-01T15:33:34Z" level=info msg="Event(v1.ObjectReference{Kind:\"Volume\", Namespace:\"longhorn-system\", Name:\"pvc-38791991-2471-4836-9a6d-4f2356a25421\", UID:\"f3a6a76b-a930-4973-8dde-c8ff994432c7\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"35366405\", FieldPath:\"\"}): type: 'Normal' reason: 'Degraded' volume pvc-38791991-2471-4836-9a6d-4f2356a25421 became degraded"
time="2023-06-01T15:33:34Z" level=debug msg="Snapshot 19ff6078-ab9f-5590-9c12-9e3e25e1ff56 with labels map[longhorn.io/for-cloning-volume:pvc-38791991-2471-4836-9a6d-4f2356a25421] for volume pvc-5c5c2f22-8f1d-4e92-89af-dd40b25fc438 already exists"
time="2023-06-01T15:33:34Z" level=info msg="Event(v1.ObjectReference{Kind:\"Volume\", Namespace:\"longhorn-system\", Name:\"pvc-38791991-2471-4836-9a6d-4f2356a25421\", UID:\"f3a6a76b-a930-4973-8dde-c8ff994432c7\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"35366405\", FieldPath:\"\"}): type: 'Normal' reason: 'VolumeCloneInitiated' source volume pvc-5c5c2f22-8f1d-4e92-89af-dd40b25fc438, snapshot 19ff6078-ab9f-5590-9c12-9e3e25e1ff56"
```

### CODE REVIEW

**Comment by ejweber** [comment]
https://ci.longhorn.io/job/private/job/longhorn-tests-regression/3974/testReport/tests/test_cloning/

File: util/util.go [hunk 1]
```diff
@@ -170,6 +172,12 @@ func RandomID() string {
 	return UUID()[:RandomIDLenth]
 }
 
+// DeterministicUUID returns a string representation of a version 5 UUID based on the provided string.
+func DeterministicUUID(data string) string {
```
Comment (PhanLe1010) [inline util/util.go:176]: Could you provide one example of the output of this function? Thank you!

**Review by PhanLe1010 — APPROVED** [review]
LGTM

**Review by innobead — APPROVED** [review]
LGTM. 

Do we have any operations similar to this case? backup?

**Comment by innobead** [comment]
@mergify backport v1.4.x

**Pull request merged**